File: design/kotlin/AgicashTokens.kt
// AgicashTokens.kt
//
// ILLUSTRATIVE consumer of design/tokens.json for Jetpack Compose (Android).
// This file is documentation of intent — not yet wired into the Android scaffold.
// When the Android app adopts it:
//   1. Copy the Kode Mono + Teko TTFs to res/font/ with snake-case names
//      (kode_mono_regular.ttf, teko_bold.ttf, etc.). See ../FONTS.md.
//   2. Optionally create font-family XML resources at res/font/kode_mono.xml
//      and res/font/teko.xml so Compose can resolve weights automatically.
//   3. Currency themes (USD/BTC) compose with light/dark via the AgicashTheme
//      composable below — dark wins over the currency theme to mirror the
//      CSS cascade in app/tailwind.css.
//
// Source of truth: ../tokens.json
// Citations: ../SOURCES.md

package com.makeprisms.agicash.design

import androidx.compose.animation.core.CubicBezierEasing
import androidx.compose.animation.core.EaseIn
import androidx.compose.animation.core.EaseInOut
import androidx.compose.animation.core.EaseOut
import androidx.compose.animation.core.LinearEasing
import androidx.compose.animation.core.tween
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Typography
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.text.font.Font
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// NOTE: R.font.* references below assume the font files have been added to
// res/font/. They are commented to keep this file syntactically valid even
// when the resources don't yet exist. Uncomment after copying the TTFs.

// -----------------------------------------------------------------------------
// Colors
// -----------------------------------------------------------------------------

object AgicashColors {
    // Light (root)
    object Light {
        val Background           = Color(0xFFFFFFFF)
        val Foreground           = Color(0xFF0A0A0A)  // hsl(0 0% 3.9%)
        val Card                 = Color(0xFFFFFFFF)
        val CardForeground       = Color(0xFFFAFAFA)
        val Popover              = Color(0xFFFFFFFF)
        val PopoverForeground    = Color(0xFF0A0A0A)
        val Primary              = Color(0xFF171717)  // hsl(0 0% 9%)
        val PrimaryForeground    = Color(0xFFFAFAFA)
        val Secondary            = Color(0xFFF5F5F5)  // hsl(0 0% 96.1%)
        val SecondaryForeground  = Color(0xFF171717)
        val Muted                = Color(0xFFF5F5F5)
        val MutedForeground      = Color(0xFF737373)  // hsl(0 0% 45.1%)
        val Accent               = Color(0xFFF5F5F5)
        val AccentForeground     = Color(0xFF171717)
        val Destructive          = Color(0xFFEF4444)  // hsl(0 84.2% 60.2%) approx
        val DestructiveForeground = Color(0xFFFAFAFA)
        val Border               = Color(0xFFE5E5E5)  // hsl(0 0% 89.8%)
        val Input                = Color(0xFFE5E5E5)
        val Ring                 = Color(0xFFD4D4D4)  // hsl(0 0% 83.1%)
    }

    // Dark
    object Dark {
        val Background           = Color(0xFF0A0A0A)
        val Foreground           = Color(0xFFFAFAFA)
        val Card                 = Color(0xFF0A0A0A)
        val CardForeground       = Color(0xFFFAFAFA)
        val Popover              = Color(0xFF0A0A0A)
        val PopoverForeground    = Color(0xFFFAFAFA)
        val Primary              = Color(0xFF1F262A)  // hsl(202 13% 13%) approx
        val PrimaryForeground    = Color(0xFFFAFAFA)
        val Secondary            = Color(0xFF262626)
        val SecondaryForeground  = Color(0xFFFAFAFA)
        val Muted                = Color(0xFF1F1F1F)
        val MutedForeground      = Color(0xFFA3A3A3)
        val Accent               = Color(0xFF262626)
        val AccentForeground     = Color(0xFFFAFAFA)
        val Destructive          = Color(0xFF7F1D1D)  // hsl(0 62.8% 30.6%) approx
        val DestructiveForeground = Color(0xFFFAFAFA)
        val Border               = Color(0xFF262626)
        val Input                = Color(0xFF262626)
        val Ring                 = Color(0xFFD4D4D4)
    }

    // USD currency theme (overrides on top of Light)
    object USD {
        val Background          = Color(0xFF004D4D)   // hsl(178 100% 15%) approx
        val Foreground          = Color(0xFFD6E5E5)
        val Primary             = Color(0xFF265C5C)
        val PrimaryForeground   = Color(0xFFD6E5E5)
        val Muted               = Color(0xFF004848)
        val MutedForeground     = Color(0xFFB8CFCF)
        val Border              = Color(0xFF006B6B)
        val Card                = Color(0xFF004848)
    }

    // BTC currency theme (overrides on top of Light)
    object BTC {
        val Background          = Color(0xFF2B5599)   // hsl(217 68% 35%) approx
        val Foreground          = Color(0xFFD6DEEB)
        val Primary             = Color(0xFF4166A6)
        val PrimaryForeground   = Color(0xFFD6DEEB)
        val Muted               = Color(0xFF305DA3)
        val MutedForeground     = Color(0xFFC2CCDE)
        val Border              = Color(0xFF4F77B8)
        val Card                = Color(0xFF305DA3)
    }
}

// -----------------------------------------------------------------------------
// Typography
// -----------------------------------------------------------------------------

// Font families — uncomment after adding TTFs to res/font/.
val AgicashKodeMono: FontFamily = FontFamily.Monospace
//val AgicashKodeMono: FontFamily = FontFamily(
//    Font(R.font.kode_mono_regular,  FontWeight.Normal),
//    Font(R.font.kode_mono_medium,   FontWeight.Medium),
//    Font(R.font.kode_mono_semibold, FontWeight.SemiBold),
//    Font(R.font.kode_mono_bold,     FontWeight.Bold),
//)

val AgicashTeko: FontFamily = FontFamily.SansSerif
//val AgicashTeko: FontFamily = FontFamily(
//    Font(R.font.teko_light,     FontWeight.Light),
//    Font(R.font.teko_regular,   FontWeight.Normal),
//    Font(R.font.teko_medium,    FontWeight.Medium),
//    Font(R.font.teko_semibold,  FontWeight.SemiBold),
//    Font(R.font.teko_bold,      FontWeight.Bold),
//)

/// Material 3 Typography mapped from the web's Tailwind scale.
/// displayLarge..titleLarge are the monetary-amount styles (Teko).
/// bodyLarge..labelSmall are the general UI styles (Kode Mono).
val AgicashTypography = Typography(
    // --- Monetary amounts (Teko) ---
    displayLarge  = TextStyle(fontFamily = AgicashTeko, fontWeight = FontWeight.Bold,     fontSize = 60.sp, lineHeight = 60.sp), // amountLG / text-6xl
    displayMedium = TextStyle(fontFamily = AgicashTeko, fontWeight = FontWeight.Bold,     fontSize = 48.sp, lineHeight = 48.sp), // amountMD / text-5xl
    displaySmall  = TextStyle(fontFamily = AgicashTeko, fontWeight = FontWeight.SemiBold, fontSize = 24.sp, lineHeight = 32.sp), // amountSM / text-2xl
    headlineLarge = TextStyle(fontFamily = AgicashTeko, fontWeight = FontWeight.SemiBold, fontSize = 20.sp, lineHeight = 28.sp), // amountXS / text-xl

    // --- General UI (Kode Mono) ---
    headlineMedium = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.SemiBold, fontSize = 30.sp, lineHeight = 36.sp),
    headlineSmall  = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.SemiBold, fontSize = 24.sp, lineHeight = 32.sp),
    titleLarge     = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.SemiBold, fontSize = 18.sp, lineHeight = 28.sp),
    titleMedium    = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Medium,   fontSize = 16.sp, lineHeight = 24.sp),
    titleSmall     = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Medium,   fontSize = 14.sp, lineHeight = 20.sp),
    bodyLarge      = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Normal,   fontSize = 16.sp, lineHeight = 24.sp),
    bodyMedium     = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Normal,   fontSize = 14.sp, lineHeight = 20.sp),
    bodySmall      = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Normal,   fontSize = 12.sp, lineHeight = 16.sp),
    labelLarge     = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Medium,   fontSize = 14.sp, lineHeight = 20.sp),
    labelMedium    = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Medium,   fontSize = 12.sp, lineHeight = 16.sp),
    labelSmall     = TextStyle(fontFamily = AgicashKodeMono, fontWeight = FontWeight.Medium,   fontSize = 10.sp, lineHeight = 16.sp), // text-2xs custom
)

// -----------------------------------------------------------------------------
// Shape
// -----------------------------------------------------------------------------

object AgicashShapes {
    val Xs        = RoundedCornerShape(2.dp)
    val Sm        = RoundedCornerShape(4.dp)
    val Md        = RoundedCornerShape(6.dp)
    val Lg        = RoundedCornerShape(8.dp)   // --radius from app/tailwind.css
    val Xl        = RoundedCornerShape(12.dp)
    val Xl2       = RoundedCornerShape(16.dp)
    val DrawerTop = RoundedCornerShape(topStart = 10.dp, topEnd = 10.dp)
    val Full      = RoundedCornerShape(9999.dp)
}

// -----------------------------------------------------------------------------
// Spacing
// -----------------------------------------------------------------------------

object AgicashSpacing {
    val Px   = 1.dp
    val S0_5 = 2.dp
    val S1   = 4.dp
    val S2   = 8.dp
    val S3   = 12.dp
    val S4   = 16.dp
    val S6   = 24.dp
    val S8   = 32.dp
    val S12  = 48.dp
}

// -----------------------------------------------------------------------------
// Motion
// -----------------------------------------------------------------------------

object AgicashMotion {
    // Durations (ms)
    const val DurationFast: Int           = 150  // transition-colors default
    const val DurationViewTransition: Int = 180  // page nav
    const val DurationBase: Int           = 200  // dialog, shake
    const val DurationSlideOutUp: Int     = 300  // toast dismiss
    const val DurationSlam: Int           = 400  // numeric slam

    // Easings — Compose ships exact CSS cubic-bezier equivalents where helpful.
    // The named constants `EaseIn`, `EaseOut`, `EaseInOut` use Google Material's
    // cubic-bezier controls. For exact CSS parity, prefer the explicit
    // CubicBezierEasing values below.
    val EaseInCss     = CubicBezierEasing(0.4f, 0f, 1f, 1f)
    val EaseOutCss    = CubicBezierEasing(0f, 0f, 0.2f, 1f)
    val EaseInOutCss  = CubicBezierEasing(0.4f, 0f, 0.2f, 1f)

    // Convenience tween factories
    fun fastTween()           = tween<Float>(DurationFast,           easing = EaseInOutCss)
    fun viewTransitionTween() = tween<Float>(DurationViewTransition, easing = EaseInCss)
    fun baseTween()           = tween<Float>(DurationBase,           easing = EaseOutCss)
    fun slideOutUpTween()     = tween<Float>(DurationSlideOutUp,     easing = EaseOutCss)
    fun slamTween()           = tween<Float>(DurationSlam,           easing = EaseOutCss)

    // No spring physics in the web app. For Compose bottom-sheets prefer the
    // platform-native ModalBottomSheet over hand-tuned springs.
}

// -----------------------------------------------------------------------------
// Theme composition
// -----------------------------------------------------------------------------

enum class CurrencyTheme { USD, BTC }
enum class AgicashColorMode { LIGHT, DARK }

data class AgicashPalette(
    val background: Color,
    val foreground: Color,
    val primary: Color,
    val primaryForeground: Color,
    val muted: Color,
    val mutedForeground: Color,
    val border: Color,
    val card: Color,
)

val LocalAgicashPalette = staticCompositionLocalOf {
    AgicashPalette(
        background = AgicashColors.Light.Background,
        foreground = AgicashColors.Light.Foreground,
        primary = AgicashColors.Light.Primary,
        primaryForeground = AgicashColors.Light.PrimaryForeground,
        muted = AgicashColors.Light.Muted,
        mutedForeground = AgicashColors.Light.MutedForeground,
        border = AgicashColors.Light.Border,
        card = AgicashColors.Light.Card,
    )
}

/**
 * Resolves the active palette, mirroring the CSS cascade in app/tailwind.css:
 * dark wins over currency theme.
 */
fun resolvePalette(currency: CurrencyTheme, mode: AgicashColorMode): AgicashPalette {
    if (mode == AgicashColorMode.DARK) {
        return AgicashPalette(
            background = AgicashColors.Dark.Background,
            foreground = AgicashColors.Dark.Foreground,
            primary = AgicashColors.Dark.Primary,
            primaryForeground = AgicashColors.Dark.PrimaryForeground,
            muted = AgicashColors.Dark.Muted,
            mutedForeground = AgicashColors.Dark.MutedForeground,
            border = AgicashColors.Dark.Border,
            card = AgicashColors.Dark.Card,
        )
    }
    return when (currency) {
        CurrencyTheme.USD -> AgicashPalette(
            background = AgicashColors.USD.Background,
            foreground = AgicashColors.USD.Foreground,
            primary = AgicashColors.USD.Primary,
            primaryForeground = AgicashColors.USD.PrimaryForeground,
            muted = AgicashColors.USD.Muted,
            mutedForeground = AgicashColors.USD.MutedForeground,
            border = AgicashColors.USD.Border,
            card = AgicashColors.USD.Card,
        )
        CurrencyTheme.BTC -> AgicashPalette(
            background = AgicashColors.BTC.Background,
            foreground = AgicashColors.BTC.Foreground,
            primary = AgicashColors.BTC.Primary,
            primaryForeground = AgicashColors.BTC.PrimaryForeground,
            muted = AgicashColors.BTC.Muted,
            mutedForeground = AgicashColors.BTC.MutedForeground,
            border = AgicashColors.BTC.Border,
            card = AgicashColors.BTC.Card,
        )
    }
}

@Composable
fun AgicashTheme(
    currency: CurrencyTheme,
    mode: AgicashColorMode,
    content: @Composable () -> Unit,
) {
    val palette = resolvePalette(currency, mode)
    CompositionLocalProvider(LocalAgicashPalette provides palette) {
        MaterialTheme(
            typography = AgicashTypography,
            content = content,
        )
    }
}

File: design/swift/AgicashTokens.swift
// AgicashTokens.swift
//
// ILLUSTRATIVE consumer of design/tokens.json for SwiftUI (iOS).
// This file is documentation of intent — not yet wired into the iOS scaffold.
// When the iOS app adopts it:
//   1. Generate Assets.xcassets Color Sets (light/dark variants) for every entry
//      in AgicashColors below. Names in Assets.xcassets must match the literals
//      passed to Color("...") here.
//   2. Bundle the Kode Mono + Teko TTFs and register them in Info.plist under
//      UIAppFonts (see ../FONTS.md for the full list).
//   3. Currency themes (USD/BTC) replace a subset of the base palette. SwiftUI
//      doesn't have a CSS-style cascade, so the recommended approach is to
//      store the active currency theme + color mode in an @EnvironmentObject
//      and resolve the right Color manually (see resolved(for:colorScheme:)).
//
// Source of truth: ../tokens.json
// Citations: ../SOURCES.md

import SwiftUI

// MARK: - Colors

/// Semantic color tokens. Each name matches an entry in Assets.xcassets, where
/// the light/dark variants are configured. The currency themes (USD/BTC) are
/// applied on top by `AgicashTheme.resolvedColor`.
enum AgicashColors {
    // Base palette — backed by Color Sets with "Any/Dark" appearances
    static let background           = Color("Background")
    static let foreground           = Color("Foreground")
    static let card                 = Color("Card")
    static let cardForeground       = Color("CardForeground")
    static let popover              = Color("Popover")
    static let popoverForeground    = Color("PopoverForeground")
    static let primary              = Color("Primary")
    static let primaryForeground    = Color("PrimaryForeground")
    static let secondary            = Color("Secondary")
    static let secondaryForeground  = Color("SecondaryForeground")
    static let muted                = Color("Muted")
    static let mutedForeground      = Color("MutedForeground")
    static let accent               = Color("Accent")
    static let accentForeground     = Color("AccentForeground")
    static let destructive          = Color("Destructive")
    static let destructiveForeground = Color("DestructiveForeground")
    static let border               = Color("Border")
    static let input                = Color("Input")
    static let ring                 = Color("Ring")

    // Hard-coded fallbacks if Assets.xcassets isn't populated yet.
    // Values pulled verbatim from app/tailwind.css (HSL → SwiftUI Color).
    enum Light {
        static let background = Color(hue: 0,    saturation: 0,    brightness: 1.0)
        static let foreground = Color(hue: 0,    saturation: 0,    brightness: 0.039)
        static let primary    = Color(hue: 0,    saturation: 0,    brightness: 0.09)
        static let destructive = Color(hue: 0/360.0, saturation: 0.842, brightness: 0.602)
    }
    enum Dark {
        static let background = Color(hue: 0, saturation: 0, brightness: 0.039)
        static let foreground = Color(hue: 0, saturation: 0, brightness: 0.98)
        static let primary    = Color(hue: 202/360.0, saturation: 0.13, brightness: 0.13)
    }
    enum USD {
        // hsl(178 100% 15%)
        static let background = Color(hue: 178/360.0, saturation: 1.0, brightness: 0.15)
        // hsl(177 42% 26%)
        static let primary    = Color(hue: 177/360.0, saturation: 0.42, brightness: 0.26)
    }
    enum BTC {
        // hsl(217 68% 35%)
        static let background = Color(hue: 217/360.0, saturation: 0.68, brightness: 0.35)
        // hsl(219 44% 45%)
        static let primary    = Color(hue: 219/360.0, saturation: 0.44, brightness: 0.45)
    }
}

// MARK: - Typography

/// Type tokens. Sizes are in points (web rem * 16). Fonts are PostScript names —
/// these only resolve if Kode Mono + Teko TTFs are bundled and registered in
/// Info.plist's UIAppFonts.
enum AgicashFont {
    // Numeric (Teko) — for monetary amounts only
    static let amountXS = Font.custom("Teko-SemiBold", size: 20)  // text-xl
    static let amountSM = Font.custom("Teko-SemiBold", size: 24)  // text-2xl
    static let amountMD = Font.custom("Teko-Bold",     size: 48)  // text-5xl
    static let amountLG = Font.custom("Teko-Bold",     size: 60)  // text-6xl

    // Primary (Kode Mono) — for all other UI text
    static let textXS   = Font.custom("KodeMono-Regular",  size: 12)
    static let textSM   = Font.custom("KodeMono-Regular",  size: 14)
    static let textBase = Font.custom("KodeMono-Regular",  size: 16)
    static let textLG   = Font.custom("KodeMono-Medium",   size: 18)
    static let text2XL  = Font.custom("KodeMono-SemiBold", size: 24)

    // Honor Dynamic Type by composing with `.scaledFont(for:)` from a custom
    // ViewModifier, or by adopting `Font.custom(_:size:relativeTo:)`:
    static func textBase(relativeTo style: Font.TextStyle) -> Font {
        Font.custom("KodeMono-Regular", size: 16, relativeTo: style)
    }
}

// MARK: - Radius

enum AgicashRadius {
    static let xs:  CGFloat = 2
    static let sm:  CGFloat = 4
    static let md:  CGFloat = 6
    static let lg:  CGFloat = 8   // matches --radius from app/tailwind.css
    static let xl:  CGFloat = 12
    static let xl2: CGFloat = 16
    static let drawerTop: CGFloat = 10
    static let full: CGFloat = 9999
}

// MARK: - Spacing

/// Spacing scale follows Tailwind's 4-pt grid. Use these instead of magic numbers
/// so the iOS layout maintains parity with the web grid.
enum AgicashSpacing {
    static let px:   CGFloat = 1
    static let s0_5: CGFloat = 2
    static let s1:   CGFloat = 4
    static let s2:   CGFloat = 8
    static let s3:   CGFloat = 12
    static let s4:   CGFloat = 16
    static let s6:   CGFloat = 24
    static let s8:   CGFloat = 32
    static let s12:  CGFloat = 48
}

// MARK: - Motion

/// Named motion tokens.
///
/// Apple's `.easeIn`/`.easeOut` are close to but not identical to CSS's named
/// curves. For pixel-level parity with the web, prefer the `.timingCurve(...)`
/// variants below; for "feels the same" parity, `.easeIn`/`.easeOut` are fine.
enum AgicashMotion {
    // Durations (seconds)
    static let durationFast:           Double = 0.15  // transition-colors default
    static let durationViewTransition: Double = 0.18  // page nav slide/fade
    static let durationBase:           Double = 0.20  // dialog open/close, shake
    static let durationSlideOutUp:     Double = 0.30  // toast dismiss
    static let durationSlam:           Double = 0.40  // numeric slam feedback

    // Named easings — exact CSS cubic-bezier control points
    static let easeIn     = Animation.timingCurve(0.4, 0, 1, 1, duration: durationBase)
    static let easeOut    = Animation.timingCurve(0, 0, 0.2, 1, duration: durationBase)
    static let easeInOut  = Animation.timingCurve(0.4, 0, 0.2, 1, duration: durationBase)

    // Apple-native equivalents — slightly softer, but feel native
    static let easeOutFast = Animation.easeOut(duration: durationFast)
    static let easeInOutBase = Animation.easeInOut(duration: durationBase)

    // No spring physics in the web app. If a SwiftUI motion needs a spring
    // (e.g., for sheet pull-down to mirror vaul's drawer), use the iOS-native
    // .sheet — it already has Apple-tuned physics that will feel more right
    // than reverse-engineering vaul's curve.
}

// MARK: - Theme resolution

enum CurrencyTheme: String { case usd, btc }
enum AgicashColorMode { case light, dark }

/// Helper to mirror the CSS cascade: dark mode wins over currency theme, exactly
/// as in app/tailwind.css where `.dark` rules appear after `.usd`/`.btc`.
struct AgicashTheme {
    let currency: CurrencyTheme
    let mode: AgicashColorMode

    var background: Color {
        switch (mode, currency) {
        case (.dark, _):       return AgicashColors.Dark.background
        case (.light, .usd):   return AgicashColors.USD.background
        case (.light, .btc):   return AgicashColors.BTC.background
        }
    }

    var primary: Color {
        switch (mode, currency) {
        case (.dark, _):       return AgicashColors.Dark.primary
        case (.light, .usd):   return AgicashColors.USD.primary
        case (.light, .btc):   return AgicashColors.BTC.primary
        }
    }
}
